### ISSUE TITLE
Protocols with read-only properties cannot be assigned objects with read/write properties

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#2371]
**Describe the bug**
Given the following code:

```python
from __future__ import annotations

from typing import Protocol


class A(Protocol):
    @property
    def name(self) -> str:
        ...


class B:
    name: str


def take_a(a: A) -> None:
    print(a)


def take_b(b: B) -> None:
    print(b)


def do_something(a: A, b: B) -> None:
    take_a(b)  # "str" is incompatible with "property"
    take_b(a)  # "A" is incompatible with "B"
```

The `take_a()` function should not generate an error.

**Expected behavior**
An error should only be generated for the invocation of `take_b()`.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool?
Command-line

Which version?
1.1.174

**Comment by erictraut** [comment]
This behavior is intended. A read/write property has semantics that are similar to an attribute, but they are not the same. If you access them from the class, you will get very different behaviors. Plus, a property typically doesn't support `del`, but it can if it provides a deleter. 

If your protocol indicates an attribute, compatible classes should also use an attribute. If your protocol indicates a property, compatible classes should use a property.

**Comment by bryanforbes** [comment]
This behavior seems to go against [PEP-544](https://www.python.org/dev/peps/pep-0544/#protocol-members):

> To distinguish between protocol class variables and protocol instance variables, the special `ClassVar` annotation should be used as specified by [PEP 526](https://www.python.org/dev/peps/pep-0526). By default, protocol variables as defined above are considered readable and writable. To define a read-only protocol variable, one can use an (abstract) property.

**Comment by erictraut** [comment]
I don't think that statement in PEP 544 contradicts our position on the compatibility of properties and attributes. It's saying that class variables within a protocol are always assumed to be read/write and that you need to use a property in your protocol if you want to enforce that they are read-only.

From a type safety standpoint, allowing a class variable of type `str` to be compatible with a read-only property that returns `str` has a couple of problems. First, it violates the read-only attribute specified in the protocol. Second, even if the property in the protocol were to include a setter, it violates other semantics of the protocol such as when accessing the symbol through the class itself.

Consider the following:

```python
class A1(Protocol):
    a: ClassVar[str]

class A2(Protocol):
    @property
    def a(self) -> str: ...

    @a.setter
    def a(self) -> str: ...

def func(a1: Type[A1], a2: Type[A2]):
    reveal_type(a1.a)  # str
    reveal_type(a2.a)  # property
```

**Comment by bryanforbes** [comment]
What concerns me about `pyright`'s behavior is that `mypy` has taken the interpretation I have described and several stubs have been implemented with that behavior in mind (`sqlalchemy-stubs` being one).

**Comment by erictraut** [comment]
In general, mypy does not do a good job tracking the difference between class variables and instance variables, so it's very "loose" when it comes to checks that involve that difference.

Can you provide a concrete example of where an existing stub has made this assumption? Is it feasible for the stub to be fixed so it works with both interpretations?

**Comment by bryanforbes** [comment]
> Can you provide a concrete example of where an existing stub has made this assumption?

Two concrete examples I can think of are from `typeshed`'s [Werkzeug](https://github.com/python/typeshed/blob/ce11072dbeda014a1702f929770628f247f5eabc/stubs/Werkzeug/werkzeug/exceptions.pyi#L7) stubs as well as [`sqlalchemy2-stubs`](https://github.com/sqlalchemy/sqlalchemy2-stubs/blob/46318625938a208c6300dc58a39ee33144517234/sqlalchemy-stubs/sql/elements.pyi#L558) using this assumption to allow passing in objects that have a readable `type` attribute (sometimes these are declared on the classes with a property and sometimes they are simply attributes, depending on the class).

> Is it feasible for the stub to be fixed so it works with both interpretations?

It seems that the only solution to make pyright see an attribute as read-only would be to define two protocols, one with an attribute and one with a read-only property, and use a union everywhere:

```python
class _Identifiable(Protocol):
    id: int

class _IdentifiableReadOnly(Protocol):
    @property
    def id(self) -> int:
        ...

Identifiable = Union[_Identifiable, _IdentifiableReadOnly]
```

This seems really verbose and prone to mistakes (if I have many read-only attributes, I would need to update them in both protocols) compared to what PEP-544 seems to suggest and [the typing community](https://github.com/python/typing/discussions/903) explicitly suggests.

**Comment by erictraut** [comment]
And you're saying that it's not feasible to change your classes to use a property rather than an attribute in these cases?

Do you understand my point about the difference in semantics between properties and attributes? I just want to make sure that's clear. If so, don't you see that as a problem with the approach that mypy has chosen to take here?

**Comment by bryanforbes** [comment]
> And you're saying that it's not feasible to change your classes to use a property rather than an attribute in these cases?

Feasibility of my classes isn't the problem since I'm writing a library. As a library author, I want to define the general shape of objects and how they will be used instead of requiring my consumers to abide by a specific class hierarchy or descriptor type which also decouples the objects within my library from each other and allows isolated testing. This is why `Protocols` and structural subtyping is so appealing: I only need to know the shapes of objects.

> Do you understand my point about the difference in semantics between properties and attributes?

I do.

> If so, don't you see that as a problem with the approach that mypy has chosen to take here?

Somewhat, but what you've pointed out seems to be a deficiency with PEP-544 that doesn't handle your edge cases very well. Further more, it seems that if the [typing community](https://github.com/python/typing/discussions/903) is suggesting the same approach that `mypy` implements, `pyright` should have that behavior as well.

In the end, all I'm looking for is how to type my code so that I can accept an object and read the value from `object.id` and guarantee to the user that `id` won't be set:

```python
class One:
    id: int

    def __init__(self, id: int) -> None:
        self.id = id

class Two:
    def __init__(self, id: int) -> None:
        self.__id = id

    @property
    def id(self) -> int:
        return self.__id

class Three:
    __slots__ = ('__data', '_id')

    def __init__(self, data: dict[str, Any]) -> None:
        self.__data = data

    @cached_slot_property('_id')
    def id(self) -> int:
        return self.__data.get('id', -1)

class Identifiable(Protocol):
    # how to type `id` so it matches all three above or other custom descriptors?

def print_id(obj: Identifiable):
    print(obj.id)
    # obj.id = 2 will produce a pyright error

one = One(1)
two = Two(2)
three = Three({'id': 3})

print_id(one)  # OK
print_id(two)  # OK
print_id(three)  # OK
```

**Comment by erictraut** [comment]
Thanks, that additional context is helpful. Let me discuss this with my colleagues, and we'll get back to you shortly.

**Comment by erictraut** [comment]
We had a good discussion about this internally. Here's what we concluded:

We'll change the behavior of pyright in the case of protocol type matching to interpret a property within the protocol to be compatible with an attribute.

This creates an ambiguity because it's not clear whether that property is meant to represent an actual property or an attribute. This becomes an issue if the member is accessed from the protocol class (as opposed to an instance of that class). To avoid this ambiguous edge case, pyright will flag it as an error if you attempt to access the property from the protocol class.

```python
class A(Protocol):
    @property
    def name(self) -> str:
        ...

def func(x: Type[A], y: A):
     x.name # Error
     y.name # Allowed
```

I think that accommodates your needs while also addressing my primary concerns.

**Comment by erictraut** [comment]
This will be included in the next release.
